test/services/katello/pulp3/repository/file/update_remote_test.rb

```diff
@@ -20,7 +20,7 @@ def setup
           @file_repo_service = @file_repo.backend_service(@mock_smart_proxy)
           @file_repo.root.update(url: 'my-files.org')
           @file_repo_service.stubs(:api).returns(@mock_api_wrapper)
-          @mock_api_wrapper.stubs(:remotes_api).returns(@mock_pulp3_api)
+          @mock_api_wrapper.stubs(:get_remotes_api).returns(@mock_pulp3_api)
```

Comment:
Perhaps this entire method shouldn't be stubbed out so it gets tested? `remotes_api` should still be stub-able.

Maybe something like  `blah_repo_service.expects(:remotes_api).once.returns(@mock_pulp3_api)`

I realize in this file only `File` repository types are tested. A new test for yum/uln could be in order :)

Reply:
If I allow it to run `get_remotes_api` (revert this change), I get:
```
[vagrant@centos8-katello-devel-stable2 foreman]$ ktest ../katello/test/services/katello/pulp3/repository/file/update_remote_test.rb
/home/vagrant/foreman/.vendor/ruby/2.7.0/gems/activerecord-6.1.7.7/lib/active_record/connection_adapters/postgresql_adapter.rb:883: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/vagrant/foreman/.vendor/ruby/2.7.0/gems/pg-1.5.6/lib/pg/text_decoder/timestamp.rb:8: warning: The called method `initialize' is defined here
Emptying /home/vagrant/foreman/jenkins/reports/unit
-- execute("SET CONSTRAINTS ALL DEFERRED;")
WARNING:  SET CONSTRAINTS can only be used in transaction blocks
   -> 0.0009s

# Running tests with run options --seed 65021:

FF.Writing XML reports to /home/vagrant/foreman/jenkins/reports/unit


Finished tests in 2.028559s, 1.4789 tests/s, 7.3944 assertions/s.


Failure:
Katello::Service::Repository::UpdateRemoteTest#test_feed_url_exists_and_remote_href_exists_updates_remote [/home/vagrant/katello/app/services/katello/pulp3/repository.rb:99]
Minitest::Assertion: unexpected invocation: #<Mock:api_wrapper>.get_remotes_api(:href => "193874298udsfsdf")
unsatisfied expectations:
- expected exactly once, invoked never: #<Mock:pulp3_api>.partial_update(any_parameters)
satisfied expectations:
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getaddresses(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getaddress(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getnames(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getname(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Katello::Pulp3::Repository>.test_remote_name(any_parameters)
- allowed any number of times, invoked never: Katello::Ping.ping(any_parameters)
- allowed any number of times, invoked never: Cert::Certs.ssl_client_key(any_parameters)
- allowed any number of times, invoked never: Cert::Certs.ssl_client_cert(any_parameters)
- allowed any number of times, invoked never: Cert::Certs.ca_cert(any_parameters)
- allowed any number of times, invoked never: #<Mock:response>.pulp_href(any_parameters)
- allowed any number of times, invoked never: #<Mock:api_wrapper>.remotes_api(any_parameters)
- allowed any number of times, invoked never: #<Mock:pulp3_api>.create(any_parameters)
- allowed any number of times, invoked never: #<Mock:smart_proxy>.pulp_primary?(any_parameters)
- allowed any number of times, invoked never: #<Mock:smart_proxy>.pulp2_preferred_for_type?(any_parameters)
- allowed any number of times, invoked once: #<Mock:smart_proxy>.pulp3_support?(any_parameters)
- allowed any number of times, invoked once: #<Katello::Pulp3::Repository::File:0x249c8>.api(any_parameters)


Failure:
Katello::Service::Repository::UpdateRemoteTest#test_feed_url_is_missing_but_remote_href_exists_deletes_remote [/home/vagrant/katello/app/services/katello/pulp3/service_common.rb:44]
Minitest::Assertion: unexpected invocation: get_remotes_api({:href=>"193874298udsfsdf"})

3 tests, 15 assertions, 2 failures, 0 errors, 0 skips

Minitest Reporters: Mean Time Report (Samples: 4, Order: :avg :desc)
Avg: 2.07335646   Min: 1.994977516  Max: 2.256206166  Last: 2.026355225  Description: Katello::Service::Repository::UpdateRemoteTest
```

Reply:
Adding entirely new tests for additional content types would make a lot of sense.

I will put it on the [agenda](https://hackmd.io/U9Q0Js0bRPez7oCgxUawng) for our upcoming meeting.

Reply:
```
- expected exactly once, invoked never: #<Mock:pulp3_api>.partial_update(any_parameters)
```
Is there a bit in your code that is expecting `partial_update` to run?

Reply:
You should be able to set it up so at least you can run get_remotes_api and stub/expect the internal method calls instead.